### PR TITLE
feat(internal): datasource defaultRegistryUrls / appendRegistryUrls

### LINKED ISSUE #5682: Refactor default registryUrls logic
If a datasource has a default registry URL (e.g. `https://registry.npmjs.org`), it should be added to the config at the datasource stage and not at the extract stage.

- [ ] Each datasource should export a new field `defaultRegistryUrls` (if applicable)
- [ ] The datasource index should check if the `registryUrls` field already has any values and if not then add the `defaultRegistryUrls` to `registryUrls` before calling the datasource's `getPkgReleases()` function
- [ ] If any datasources always need a "fallback" url or urls, we could have `appendRegistryUrls` instead of `defaultRegistryUrls` and *add* them even if the `registryUrls` field in config is non-empty
- [ ] Datasource index can maybe skip a lookup itself if the datasource has no default or append registry URLs and also the `registryUrls` in config are missing
- [ ] Remove any adding of default registry URLs from extract phase

The purpose of this is to reduce the amount of common `registryUrls` handling code in each datasource implementation.

### PR DESCRIPTION
**Pull request by zharinov**
Closes #5682

### CODE REVIEW

File: lib/datasource/maven/common.ts [hunk 1]
```diff
@@ -1,5 +1,3 @@
 export const id = 'maven';
 
 export const MAVEN_REPO = 'https://repo.maven.apache.org/maven2';
-export const MAVEN_REPO_DEPRECATED = 'https://central.maven.org/maven2';
```
Comment (zharinov) [inline lib/datasource/maven/common.ts]: https://central.maven.org/maven2 seems to be completely down

**Comment by zharinov** [comment]
I think it's enough at the moment. We're still having `pip_requirements` and `poetry` which use more complicated logic, other managers should be okay.

**Review by viceice — DISMISSED** [review]

File: lib/datasource/sbt-plugin/index.ts [hunk 2]
```diff
@@ -1,12 +1,15 @@
...
 import { downloadHttpProtocol } from '../maven/util';
 import { parseIndexDir, SBT_PLUGINS_REPO } from './util';
+import { MAVEN_REPO } from '../maven/common';
 import { logger } from '../../logger';
 import { GetReleasesConfig, ReleaseResult } from '../common';
 import { resolvePackageReleases } from '../sbt-package';
 
 export const id = 'sbt-plugin';
 
+export const defaultRegistryUrls = [SBT_PLUGINS_REPO, MAVEN_REPO];
```
Comment (rarkins) [inline lib/datasource/sbt-plugin/index.ts]: Is this a change in functionality? Or were we guaranteed that the existing `registryUrls` were always set to `MAVEN_REPO`?

File: lib/manager/leiningen/extract.ts [hunk 3]
```diff
@@ -96,7 +95,7 @@ export function extractFromVectors(
 }
 
 function extractLeinRepos(content: string): string[] {
-  const result = [CLOJARS_REPO, MAVEN_REPO];
```
Comment (rarkins) [inline lib/manager/leiningen/extract.ts]: Previously, it looks like these were prepended to any configured repositories. But now they are `defaultRegistryUrls`. This is a functional change, right?

Comment (zharinov) [reply lib/manager/leiningen/extract.ts]: It should work fine after merge, affecting only order of HTTP request

File: lib/datasource/sbt-plugin/index.ts (hunk 2 above)
Comment (zharinov) [reply lib/datasource/sbt-plugin/index.ts]: Probably, we want to add `MAVEN_REPO` here later for some plugins located in central maven. For now, I've removed it.

**Review by rarkins — CHANGES_REQUESTED** [review]
@zharinov needs de-conflicting

File: lib/datasource/index.ts [hunk 4]
```diff
@@ -25,6 +26,18 @@ function load(datasource: string): Promise<Datasource> {
   return datasources.get(datasource);
 }
 
+async function resolveRegistryUrls(
+  manager: string,
+  extractedUrls: string[]
+): Promise<string[]> {
+  const urls = extractedUrls || [];
```
Comment (viceice) [inline lib/datasource/index.ts]: Do we need this? Can't we simply pass `extractedUrls` to below

File: lib/datasource/index.ts [hunk 5]
```diff
@@ -37,7 +50,14 @@ async function fetchReleases(
...
   }
-  const dep = await (await load(datasource)).getPkgReleases(config);
+  const registryUrls = await resolveRegistryUrls(
+    datasource,
+    config.registryUrls
+  );
+  const dep = await (await load(datasource)).getPkgReleases({
+    ...config,
+    registryUrls,
+  });
```
Comment (viceice) [inline lib/datasource/index.ts]: Why not first await datasource and then passing to `resolveRegistryUrls` ?

**Review by viceice — COMMENTED** [review]
Some small things

File: lib/datasource/index.ts (hunk 5 above)
Comment (rarkins) [reply lib/datasource/index.ts]: Can you explain more? I don't follow

Comment (zharinov) [reply lib/datasource/index.ts]: Got it, I'll fix

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs de-conflicting

**Comment by rarkins** [comment]
This somehow broke nuget in production but I can't work out why.

Example repository: https://github.com/felixfbecker/PSKubectl

With renovate@19.198.1:

```
{"level":30,"stats":{"managers":{"npm":{"fileCount":1,"depCount":2},"nuget":{"fileCount":1,"depCount":4}},"fileCount":2,"depCount":6},"msg":"Extraction statistics","time":"2020-04-07T12:57:54.134Z"}
{"level":20,"dependency":"System.Management.Automation","msg":"Dependency is ignored","time":"2020-04-07T12:57:54.244Z"}
{"level":20,"packageFile":"package.json","depName":"semantic-release","msg":"fetchDepUpdates finished","time":"2020-04-07T12:57:54.425Z"}
{"level":20,"packageFile":"package.json","depName":"@semantic-release/exec","msg":"fetchDepUpdates finished","time":"2020-04-07T12:57:54.608Z"}
{"level":20,"packageFile":"src/PSKubectl.csproj","depName":"KubeClient","msg":"fetchDepUpdates finished","time":"2020-04-07T12:57:55.198Z"}
{"level":20,"packageFile":"src/PSKubectl.csproj","depName":"KubeClient.Extensions.KubeConfig","msg":"fetchDepUpdates finished","time":"2020-04-07T12:57:55.353Z"}
{"level":20,"packageFile":"src/PSKubectl.csproj","depName":"Serilog.Extensions.Logging.File","msg":"fetchDepUpdates finished","time":"2020-04-07T12:57:55.466Z"}
{"level":20,"msg":"fetchUpdates complete","time":"2020-04-07T12:57:55.466Z"}
```

With renovate@19.200.1:

```
{"level":30,"stats":{"managers":{"npm":{"fileCount":1,"depCount":2},"nuget":{"fileCount":1,"depCount":4}},"fileCount":2,"depCount":6},"msg":"Extraction statistics","time":"2020-04-07T12:33:50.570Z"}
{"level":20,"dependency":"System.Management.Automation","msg":"Dependency is ignored","time":"2020-04-07T12:33:50.626Z"}
{"level":20,"lookupName":"KubeClient","msg":"Dependency lookup failure: not found in all feeds","time":"2020-04-07T12:33:50.640Z"}
{"level":20,"lookupName":"KubeClient.Extensions.KubeConfig","msg":"Dependency lookup failure: not found in all feeds","time":"2020-04-07T12:33:50.641Z"}
{"level":20,"lookupName":"Serilog.Extensions.Logging.File","msg":"Dependency lookup failure: not found in all feeds","time":"2020-04-07T12:33:50.641Z"}
{"level":20,"dependency":"KubeClient","packageFile":"src/PSKubectl.csproj","msg":"Failed to look up dependency KubeClient","time":"2020-04-07T12:33:50.641Z"}
{"level":20,"dependency":"KubeClient.Extensions.KubeConfig","packageFile":"src/PSKubectl.csproj","msg":"Failed to look up dependency KubeClient.Extensions.KubeConfig","time":"2020-04-07T12:33:50.641Z"}
{"level":20,"dependency":"Serilog.Extensions.Logging.File","packageFile":"src/PSKubectl.csproj","msg":"Failed to look up dependency Serilog.Extensions.Logging.File","time":"2020-04-07T12:33:50.641Z"}
{"level":20,"packageFile":"src/PSKubectl.csproj","depName":"KubeClient","msg":"fetchDepUpdates finished","time":"2020-04-07T12:33:50.641Z"}
{"level":20,"packageFile":"src/PSKubectl.csproj","depName":"KubeClient.Extensions.KubeConfig","msg":"fetchDepUpdates finished","time":"2020-04-07T12:33:50.641Z"}
{"level":20,"packageFile":"src/PSKubectl.csproj","depName":"Serilog.Extensions.Logging.File","msg":"fetchDepUpdates finished","time":"2020-04-07T12:33:50.641Z"}
{"level":20,"packageFile":"package.json","depName":"semantic-release","msg":"fetchDepUpdates finished","time":"2020-04-07T12:33:50.701Z"}
{"level":20,"packageFile":"package.json","depName":"@semantic-release/exec","msg":"fetchDepUpdates finished","time":"2020-04-07T12:33:50.795Z"}
{"level":20,"msg":"fetchUpdates complete","time":"2020-04-07T12:33:50.795Z"}
```

**Comment by rarkins** [comment]
Not working:

```
{
  "level": 20,
  "config": {
    "npm": [
      {
        "packageFile": "package.json",
        "deps": [
          {
            "depType": "devDependencies",
            "depName": "@semantic-release/exec",
            "currentValue": "^5.0.0",
            "datasource": "npm",
            "prettyDepType": "devDependency",
            "lockedVersion": "5.0.0",
            "updates": [],
            "warnings": [],
            "sourceUrl": "https://github.com/semantic-release/exec"
          },
          {
            "depType": "devDependencies",
            "depName": "semantic-release",
            "currentValue": "^17.0.0",
            "datasource": "npm",
            "prettyDepType": "devDependency",
            "lockedVersion": "17.0.0",
            "updates": [],
            "warnings": [],
            "sourceUrl": "https://github.com/semantic-release/semantic-release"
          }
        ],
        "packageJsonType": "app",
        "npmLock": "package-lock.json",
        "skipInstalls": true
      }
    ],
    "nuget": [
      {
        "packageFile": "src/PSKubectl.csproj",
        "manager": "nuget",
        "deps": [
          {
            "depType": "nuget",
            "depName": "KubeClient",
            "currentValue": "2.3.9",
            "managerData": {
              "lineNumber": 8
            },
            "datasource": "nuget",
            "updates": [],
            "warnings": [
              {
                "updateType": "warning",
                "message": "Failed to look up dependency KubeClient"
              }
            ]
          },
          {
            "depType": "nuget",
            "depName": "KubeClient.Extensions.KubeConfig",
            "currentValue": "2.3.9",
            "managerData": {
              "lineNumber": 9
            },
            "datasource": "nuget",
            "updates": [],
            "warnings": [
              {
                "updateType": "warning",
                "message": "Failed to look up dependency KubeClient.Extensions.KubeConfig"
              }
            ]
          },
          {
            "depType": "nuget",
            "depName": "System.Management.Automation",
            "currentValue": "6.1.1",
            "managerData": {
              "lineNumber": 10
            },
            "datasource": "nuget",
            "updates": [],
            "skipReason": "ignored"
          },
          {
            "depType": "nuget",
            "depName": "Serilog.Extensions.Logging.File",
            "currentValue": "1.1.0",
            "managerData": {
              "lineNumber": 14
            },
            "datasource": "nuget",
            "updates": [],
            "warnings": [
              {
                "updateType": "warning",
                "message": "Failed to look up dependency Serilog.Extensions.Logging.File"
              }
            ]
          }
        ]
      }
    ]
  },
  "msg": "packageFiles with updates",
  "time": "2020-04-07T12:33:50.796Z"
}
```

Working:
```
{
  "level": 20,
  "config": {
    "npm": [
      {
        "packageFile": "package.json",
        "deps": [
          {
            "depType": "devDependencies",
            "depName": "@semantic-release/exec",
            "currentValue": "^5.0.0",
            "datasource": "npm",
            "prettyDepType": "devDependency",
            "lockedVersion": "5.0.0",
            "updates": [],
            "warnings": [],
            "sourceUrl": "https://github.com/semantic-release/exec"
          },
          {
            "depType": "devDependencies",
            "depName": "semantic-release",
            "currentValue": "^17.0.0",
            "datasource": "npm",
            "prettyDepType": "devDependency",
            "lockedVersion": "17.0.0",
            "updates": [],
            "warnings": [],
            "sourceUrl": "https://github.com/semantic-release/semantic-release"
          }
        ],
        "packageJsonType": "app",
        "npmLock": "package-lock.json",
        "skipInstalls": true
      }
    ],
    "nuget": [
      {
        "packageFile": "src/PSKubectl.csproj",
        "manager": "nuget",
        "deps": [
          {
            "depType": "nuget",
            "depName": "KubeClient",
            "currentValue": "2.3.9",
            "managerData": {
              "lineNumber": 8
            },
            "datasource": "nuget",
            "updates": [
              {
                "fromVersion": "2.3.9",
                "toVersion": "2.3.11",
                "newValue": "2.3.11",
                "newMajor": 2,
                "newMinor": 3,
                "updateType": "minor",
                "isSingleVersion": true
              }
            ],
            "warnings": [],
            "sourceUrl": "https://github.com/tintoy/dotnet-kube-client"
          },
          {
            "depType": "nuget",
            "depName": "KubeClient.Extensions.KubeConfig",
            "currentValue": "2.3.9",
            "managerData": {
              "lineNumber": 9
            },
            "datasource": "nuget",
            "updates": [
              {
                "fromVersion": "2.3.9",
                "toVersion": "2.3.11",
                "newValue": "2.3.11",
                "newMajor": 2,
                "newMinor": 3,
                "updateType": "minor",
                "isSingleVersion": true
              }
            ],
            "warnings": [],
            "sourceUrl": "https://github.com/tintoy/dotnet-kube-client"
          },
          {
            "depType": "nuget",
            "depName": "System.Management.Automation",
            "currentValue": "6.1.1",
            "managerData": {
              "lineNumber": 10
            },
            "datasource": "nuget",
            "updates": [],
            "skipReason": "ignored"
          },
          {
            "depType": "nuget",
            "depName": "Serilog.Extensions.Logging.File",
            "currentValue": "1.1.0",
            "managerData": {
              "lineNumber": 14
            },
            "datasource": "nuget",
            "updates": [
              {
                "fromVersion": "1.1.0",
                "toVersion": "2.0.0",
                "newValue": "2.0.0",
                "newMajor": 2,
                "newMinor": 0,
                "updateType": "major",
                "isSingleVersion": true
              }
            ],
            "warnings": [],
            "sourceUrl": "https://github.com/serilog/serilog-extensions-logging-file"
          }
        ]
      }
    ]
  },
  "msg": "packageFiles with updates",
  "time": "2020-04-07T12:57:55.468Z"
}
```

When I run it locally:
```
       "config": {
         "npm": [
           {
             "packageFile": "package.json",
             "deps": [
               {
                 "depType": "devDependencies",
                 "depName": "@semantic-release/exec",
                 "currentValue": "^5.0.0",
                 "datasource": "npm",
                 "prettyDepType": "devDependency",
                 "lockedVersion": "5.0.0",
                 "updates": [],
                 "warnings": [],
                 "sourceUrl": "https://github.com/semantic-release/exec"
               },
               {
                 "depType": "devDependencies",
                 "depName": "semantic-release",
                 "currentValue": "^17.0.0",
                 "datasource": "npm",
                 "prettyDepType": "devDependency",
                 "lockedVersion": "17.0.0",
                 "updates": [],
                 "warnings": [],
                 "sourceUrl": "https://github.com/semantic-release/semantic-release"
               }
             ],
             "packageJsonType": "app",
             "npmLock": "package-lock.json",
             "skipInstalls": true
           }
         ],
         "nuget": [
           {
             "packageFile": "src/PSKubectl.csproj",
             "manager": "nuget",
             "deps": [
               {
                 "depType": "nuget",
                 "depName": "KubeClient",
                 "currentValue": "2.3.9",
                 "managerData": {"lineNumber": 8},
                 "datasource": "nuget",
                 "registryUrls": [
                   "https://api.nuget.org/v3/index.json",
                   "https://www.myget.org/F/dotnet-kube-client/api/v3/index.json"
                 ],
                 "updates": [
                   {
                     "fromVersion": "2.3.9",
                     "toVersion": "2.3.11",
                     "newValue": "2.3.11",
                     "newMajor": 2,
                     "newMinor": 3,
                     "updateType": "minor",
                     "isSingleVersion": true
                   }
                 ],
                 "warnings": [],
                 "sourceUrl": "https://github.com/tintoy/dotnet-kube-client"
               },
               {
                 "depType": "nuget",
                 "depName": "KubeClient.Extensions.KubeConfig",
                 "currentValue": "2.3.9",
                 "managerData": {"lineNumber": 9},
                 "datasource": "nuget",
                 "registryUrls": [
                   "https://api.nuget.org/v3/index.json",
                   "https://www.myget.org/F/dotnet-kube-client/api/v3/index.json"
                 ],
                 "updates": [
                   {
                     "fromVersion": "2.3.9",
                     "toVersion": "2.3.11",
                     "newValue": "2.3.11",
                     "newMajor": 2,
                     "newMinor": 3,
                     "updateType": "minor",
                     "isSingleVersion": true
                   }
                 ],
                 "warnings": [],
                 "sourceUrl": "https://github.com/tintoy/dotnet-kube-client"
               },
               {
                 "depType": "nuget",
                 "depName": "System.Management.Automation",
                 "currentValue": "6.1.1",
                 "managerData": {"lineNumber": 10},
                 "datasource": "nuget",
                 "registryUrls": [
                   "https://api.nuget.org/v3/index.json",
                   "https://www.myget.org/F/dotnet-kube-client/api/v3/index.json"
                 ],
                 "updates": [],
                 "skipReason": "ignored"
               },
               {
                 "depType": "nuget",
                 "depName": "Serilog.Extensions.Logging.File",
                 "currentValue": "1.1.0",
                 "managerData": {"lineNumber": 14},
                 "datasource": "nuget",
                 "registryUrls": [
                   "https://api.nuget.org/v3/index.json",
                   "https://www.myget.org/F/dotnet-kube-client/api/v3/index.json"
                 ],
                 "updates": [
                   {
                     "fromVersion": "1.1.0",
                     "toVersion": "2.0.0",
                     "newValue": "2.0.0",
                     "newMajor": 2,
                     "newMinor": 0,
                     "updateType": "major",
                     "isSingleVersion": true
                   }
                 ],
                 "warnings": [],
                 "sourceUrl": "https://github.com/serilog/serilog-extensions-logging-file"
               }
             ]
           }
         ]
       }
``

**Comment by rarkins** [comment]
Notice that I run it locally and find `registryUrls`. Is that part perhaps case-related?

**Pull request merged**